### PR TITLE
rustc: use a separate copy of P for HIR than for AST.

### PR DESCRIPTION
**Pull request by eddyb**
Note: this currently includes/is based on top of #61987.

Like #61968, but goes one step further and uses a separate `P<...>` for the HIR, with no `Clone`, or the ability to mutate after allocation.
There is still `into_inner`/`into_iter`, but they're only exposed for `hir::lowering`, and they would take more work to untangle.

r? @petrochenkov cc @rust-lang/compiler

### CODE REVIEW

File: src/librustc/hir/mod.rs [hunk 1]
```diff
@@ -1983,9 +1984,9 @@ pub struct InlineAsmOutput {
 pub struct InlineAsm {
     pub asm: Symbol,
     pub asm_str_style: StrStyle,
-    pub outputs: HirVec<InlineAsmOutput>,
-    pub inputs: HirVec<Symbol>,
-    pub clobbers: HirVec<Symbol>,
+    pub outputs: Vec<InlineAsmOutput>,
+    pub inputs: Vec<Symbol>,
+    pub clobbers: Vec<Symbol>,
```
Comment (petrochenkov) [inline src/librustc/hir/mod.rs]: Unclear why these are vectors.
Are they mutated in HIR somehow?

File: src/librustc/hir/ptr.rs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+use rustc_data_structures::stable_hasher::{StableHasher, StableHasherResult,
+                                           HashStable};
+/// An owned smart pointer.
+#[derive(Hash, PartialEq, Eq)]
+pub struct P<T: ?Sized> {
+    ptr: Box<T>
+}
+
+#[allow(non_snake_case)]
```
Comment (petrochenkov) [inline src/librustc/hir/ptr.rs]: Style nit: attribute before doc comments.

Comment (eddyb) [reply src/librustc/hir/ptr.rs]: (I think it's pre-existing in `syntax::ptr`, I'll have to change both).

File: src/librustc/hir/mod.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc/hir/mod.rs]: It's to keep `Clone` on `InlineAsm`, I believe `mir::StatementKind::InlineAsm` contains a `hir::InlineAsm`? So in a sense this `struct` is not specific to HIR.

Comment (petrochenkov) [reply src/librustc/hir/mod.rs]: Could you add an explanation comment?

**Comment by eddyb** [comment]
@petrochenkov Rebased and fixed nits.

**Comment by pietroalbini** [comment]
@bors retry

Yielding priority to the Azure migration.

**Comment by pietroalbini** [comment]
@bors retry

Yielding priority to the stable rebuild.

**Pull request merged**